### PR TITLE
Nano : ResNet Demo for InferenceOptmizer

### PR DESCRIPTION
**Pull request by rnwang04**
## Description

Use Finetuned ResNet18 as an example to illustrate how to apply InferenceOptimizer to quickly find acceleration method with the minimum inference latency.

### 1. Why the change?

To provide a simple example illustrating how to use InferenceOptimizer.

### 2. User API changes

no changes.

### 3. Summary of the change

Basic example for usage of InferenceOptimizer.

### CODE REVIEW

File: python/nano/example/pytorch/inference_pipeline/inference_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+def accuracy(pred, target):
+    pred = torch.sigmoid(pred)
+    target = target.view((-1, 1)).type_as(pred).int()
+    return Accuracy()(pred, target)
+
+# accelaration inference using InferenceOptimizer
+model.eval()
+optimizer = InferenceOptimizer()
+optimizer.optimize(model=model,
+                   training_data=datamodule.predict_dataloader(),
```
Comment (TheaperDeng) [inline python/nano/example/pytorch/inference_pipeline/inference_pipeline.py:305]: we need to use training dataloader here to avoid data leak as we stated here https://github.com/intel-analytics/BigDL/blob/1b45bf03830cd5c47edaf23bd78705edd5143313/python/nano/src/bigdl/nano/pytorch/inference/optimizer.py#L115-L121

File: python/nano/example/pytorch/inference_pipeline/inference_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+                   validation_data=datamodule.val_dataloader(),
+                   metric=accuracy,
+                   direction="max",
+                   cpu_num=1,
+                   latency_sample_num=30)
```
Comment (TheaperDeng) [inline python/nano/example/pytorch/inference_pipeline/inference_pipeline.py:310]: how long does it take?

File: python/nano/example/pytorch/inference_pipeline/inference_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+for key, value in optimizer.optimized_model_dict.items():
+    print("accleration option: {}, latency: {:.4f}ms, accuracy: {:.4f}".format(key, value["latency"], value["accuracy"]))
+
+acc_model, option = optimizer.get_best_model(accelerator="onnxruntime")
+print(option)
+
+acc_model, option = optimizer.get_best_model(accuracy_criterion=0.05)
+print(option)
```
Comment (TheaperDeng) [inline python/nano/example/pytorch/inference_pipeline/inference_pipeline.py:319]: show how to use `acc_model` to predict.

**Comment by jason-dai** [comment]
We want to focus on the demo of inference pipeline optimization, but 90% of the code in this demo is not related to that. We need to make sure out focus area stands out from the code.

**Comment by rnwang04** [comment]
> We want to focus on the demo of inference pipeline optimization, but 90% of the code in this demo is not related to that. We need to make sure out focus area stands out from the code.

How about moving these code(model definition, dataset and loader definition, finetuning) to another _help.py file and only keep last lines about inference in this pipeline.py file ? @jason-dai @TheaperDeng

**Comment by TheaperDeng** [comment]
> > We want to focus on the demo of inference pipeline optimization, but 90% of the code in this demo is not related to that. We need to make sure out focus area stands out from the code.
> 
> How about moving these code(model definition, dataset and loader definition, finetuning) to another file(_help.py or _finetune.py) and only keep last lines about inference in this pipeline.py file ? @jason-dai @TheaperDeng

I think that would be fine.

File: python/nano/example/pytorch/inference_pipeline/resnet/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
+# Bigdl-nano InferenceOptimizer example on Cat vs. Dog dataset
+
+This example illustrates how to apply InferenceOptimizer to quickly find acceleration method with the minimum inference latency under specific restrictions or without restrictions for a trained model. 
+For the sake of this example, we first train the proposed network(by default, a ResNet50 is used) on the [cats and dogs dataset](https://storage.googleapis.com/mledu-datasets/cats_and_dogs_filtered.zip), which consists both [frozen and unfrozen stages](https://github.com/PyTorchLightning/pytorch-lightning/blob/495812878dfe2e31ec2143c071127990afbb082b/pl_examples/domain_templates/computer_vision_fine_tuning.py#L21-L35). Then, by calling `optimize()`, we can obtain all avaliable accelaration combinations provided by BigDL-Nano for inference. By calling `get_best_mdoel()` , we could get an accelerated model whose inference is 7.8x times faster.
```
Comment (TheaperDeng) [inline python/nano/example/pytorch/inference_pipeline/resnet/README.md:4]: avaliable -> available

File: python/nano/example/pytorch/inference_pipeline/resnet/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+## Prepare the environment
+We recommend you to use [Anaconda](https://www.anaconda.com/distribution/#linux) to prepare the environment.
+```
+conda create -n nano python=3.7  # "nano" is conda environment name, you can use any name you like.
+conda activate nano
+pip install jsonargparse[signatures]
+pip install --pre --upgrade bigdl-nano[pytorch]
+
+# bf16 is avaliable only on torch1.12
+pip install torch==1.12.0 torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cpu
```
Comment (TheaperDeng) [inline python/nano/example/pytorch/inference_pipeline/resnet/README.md:16]: we don't need `torchaudio`?

File: python/nano/example/pytorch/inference_pipeline/resnet/README.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+pip install --pre --upgrade bigdl-nano[pytorch]
+
+# bf16 is avaliable only on torch1.12
+pip install torch==1.12.0 torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cpu 
+# Necessary packages for inference accelaration
+pip install --upgrade intel-extension-for-pytorch
+pip install onnx onnxruntime onnxruntime-extensions
+pip install openvino-dev
+pip install --upgrade neural-compressor
+```
```
Comment (TheaperDeng) [inline python/nano/example/pytorch/inference_pipeline/resnet/README.md:22]: may be we should tell our users to ignore the warnings might pop out here? (about numpy version)

**Review by TheaperDeng — COMMENTED** [review]
others LGTM

File: python/nano/example/pytorch/inference_pipeline/resnet/README.md (hunk 5 above)
Comment (rnwang04) [reply python/nano/example/pytorch/inference_pipeline/resnet/README.md:16]: Yes, good catch. thanks : ) Done.

File: python/nano/example/pytorch/inference_pipeline/resnet/README.md (hunk 4 above)
Comment (rnwang04) [reply python/nano/example/pytorch/inference_pipeline/resnet/README.md:4]: Done

File: python/nano/example/pytorch/inference_pipeline/resnet/README.md (hunk 6 above)
Comment (rnwang04) [reply python/nano/example/pytorch/inference_pipeline/resnet/README.md:24]: done

**Review by TheaperDeng — APPROVED** [review]

**Comment by rnwang04** [comment]
# Update 9/1
To make this demo run faster, there are some modifications:
- replace Resnet50 with Resnet18
- Only finetune 2 epochs
- Use subset of validation dataset(80 samples) to calculate accuracy
 
Now, fitune cost about 2 min and inference cost about 4 min (on cpx1 which has 112 cores). 
@jason-dai @TheaperDeng

File: python/nano/example/pytorch/inference_pipeline/resnet/inference_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    optimizer = InferenceOptimizer()
+    # optimize may take about 2 minutes to run all possible accelaration combinations
+    optimizer.optimize(model=model,
+                       # To obtain the latency of single sample, set batch_size=1
+                       training_data=datamodule.train_dataloader(batch_size=1),
+                       # here we only take part samples to calculate a rough accuracy
+                       validation_data=datamodule.val_dataloader(limit_num_samples=160),
+                       metric=accuracy,
+                       direction="max",
+                       cpu_num=1,
```
Comment (TheaperDeng) [inline python/nano/example/pytorch/inference_pipeline/resnet/inference_pipeline.py:49]: I am not sure if set this `cpu_num` to 1 is proper at least for now. Since we know there are many limitation and know issues in our resource control for onnxrt and openvino. 

Would it be more reasonable if we just leave this `cpu_num` to be None which means all cores can be used without restrictions.

**Comment by TheaperDeng** [comment]
will first merge this PR and fix any issues come up later
